# 01. Import the required libraries

In [33]:
import pandas as pd
import numpy as np
import os

# 02. Importing the data

In [42]:
# Importing orders.csv
df_ords = pd.read_csv(r'/Users/omkargopalakrishnan/Instacart Basket Analysis/Original Data/orders.csv', index_col = False)

In [35]:
# Import products.csv
df_prods = pd.read_csv(r'/Users/omkargopalakrishnan/Instacart Basket Analysis/Original Data/products.csv', index_col = False)

In [36]:
# Telling Python to remember a main folder path
path = r'/Users/omkargopalakrishnan/Instacart Basket Analysis/'

In [37]:
path

'/Users/omkargopalakrishnan/Instacart Basket Analysis/'

In [38]:
# Simplify the import function
df_dep = pd.read_csv(os.path.join(path, 'Original Data', 'orders.csv'), index_col = False)

# Mixed Type Data

In [3]:
# Creates a new dataframe called df_test
df_test = pd.DataFrame()

In [4]:
# Create a new column, mix, within df_test and fills it with numeric, string, and boolean values
df_test['mix'] = ['a', 'b', 1, True]

In [5]:
# Check the dataframe 
df_test.head 

<bound method NDFrame.head of     mix
0     a
1     b
2     1
3  True>

In [6]:
# Checking whether the dataframe contains any mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [7]:
# Convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

# Missing Values

In [14]:
# Assign the function isnull() to the df_prods dataframe, then sum the result with the attached sum() function
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name' ] .isnull () == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape # The shape function returns the number of rows and columns in a dataframe

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False] # create a new dataframe called df_prods_clean

In [19]:
df_prods_clean.shape

(49677, 5)

In [35]:
df_prods.dropna(inplace = True) # Another way you can drop all missing values

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Duplicates

In [20]:
df_dups = df_prods_clean[df_prods_clean.duplicated()] # creates a new subset of df_prods_clean—df_dups—containing only rows that are duplicates

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [23]:
df_prods_clean.shape

(49677, 5)

In [25]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates() # create a new dataframe excluding the duplicates identified using the drop_duplicates() function. It contains only the unique rows from df_prods_clean

In [26]:
df_prods_clean_no_dups.shape

(49672, 5)

In [29]:
df_prods.isnull() # Running the `df.isnull()` function should be one of the first steps you take when working with a new dataframe. Identifying nulls is especially important in understanding the availability of your data.

product_id  product_name  aisle_id  department_id  prices
0           False         False     False          False   False
1           False         False     False          False   False
2           False         False     False          False   False
3           False         False     False          False   False
4           False         False     False          False   False
...           ...           ...       ...            ...     ...
49688       False         False     False          False   False
49689       False         False     False          False   False
49690       False         False     False          False   False
49691       False         False     False          False   False
49692       False         False     False          False   False

[49693 rows x 5 columns]

In [30]:
df_prods_clean_no_dups.isnull()

product_id  product_name  aisle_id  department_id  prices
0           False         False     False          False   False
1           False         False     False          False   False
2           False         False     False          False   False
3           False         False     False          False   False
4           False         False     False          False   False
...           ...           ...       ...            ...     ...
49688       False         False     False          False   False
49689       False         False     False          False   False
49690       False         False     False          False   False
49691       False         False     False          False   False
49692       False         False     False          False   False

[49672 rows x 5 columns]

In [40]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'products_checked.csv'))

# TASK

## 1. If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now

#### Done

## 2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. Tip: Keep an eye on min and max values!

In [31]:
df_ords.describe() # returns a summary of statistics for each numerical column in the DataFrame

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### The values in the columns 'order_number', 'orders_day_of_week', and 'days_since_prior_order'. 'order_number' have a maximum value of 100, which seems very high considering the time frame. 'orders_day_of_week' has a minimum of '0' and a maximum of '6', which indicates that the days are zero-indexed (0 for Monday or Sunday). 'days_since_prior_order' has a maximum value of 30 suggests that orders are saved monthly. Now let us look at the zero values for 'days_since_prior_order'. The minimum value for 'days_since_prior_order' is 0, suggesting multiple orders placed on the same day or a placeholder for the first orders made by the customers.  The count is lower compared to the other columns, suggesting missing data. Finally, 'order_id', 'user_id', and 'order_number' should be converted from numeric to string values.

## 3. Check for mixed-type data in your df_ords dataframe

In [105]:
# If the variable weird is False, that means the dataframe columns contain only one data type. No mixed-type data detected.
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

## 4. If you find mixed-type data, fix it. The column in question should contain observations of a single data type

In [48]:
df_ords.dtypes # check the data types in the dataframe orders

order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
order_dow                   int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

#### All variables have unique data types. No mixed data detected in the df_ords dataframe

## 5. Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find

In [49]:
df_ords.isnull().sum() # check whether there are null values

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### The 'days_since_prior_order' column has missing 206209 values. The many missing values could be because customers' first orders do not have a value. This could be relevant information that should be preserved, and not be removed.

## 6. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice

In [51]:
# In order to determine what course of action to take regarding the missing values, we must separate the data from the original dataframe
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [53]:
df_ords_nan # check for missing values (NaNs) in the DataFrame

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [54]:
# Verify whether all 'order_number' values equal to 1 are missing values in the 'days_since_prior_order' column

In [56]:
# Filter the dataframe where 'order_number' equals 1
df_first_orders = df_ords_nan[df_ords_nan['order_number'] == '1']

In [58]:
# Check if all of these have missing values in 'days_since_prior_order'
df_first_orders_missing = df_first_orders['days_since_prior_order'].isna().all()

In [61]:
# Print the result
print("All first orders have missing 'days_since_prior_order':", df_first_orders_missing) 

All first orders have missing 'days_since_prior_order': True


In [64]:
# We know for certain that every row with 'order_number' equal to 1 has a missing value in the'days_since_prior_order' column

In [66]:
# Address the missing values by creating a column that flags order numbers as either True/False depending on if the 'order_number' = 1
df_ords['first_order'] = df_ords['days_since_prior_order'].isna()

In [68]:
df_ords

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
1         2398795        1    prior             2          3   
2          473747        1    prior             3          3   
3         2254736        1    prior             4          4   
4          431534        1    prior             5          4   
...           ...      ...      ...           ...        ...   
3421078   2266710   206209    prior            10          5   
3421079   1854736   206209    prior            11          4   
3421080    626363   206209    prior            12          1   
3421081   2977660   206209    prior            13          1   
3421082    272231   206209    train            14          6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 8 columns]

In [70]:
# Check data-type of the new column and update if necessary
df_ords['first_order'].dtype

dtype('bool')

In [72]:
# Re-check data to make sure totals are accurate
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
first_order                    0
dtype: int64

In [108]:
# Missing values have been addressed appropriately, now the client can see why there are missing values in the 'order_number' column. 
# Removing/filtering these values would've been incorrect because the missing data is quite valuable.

## 7. Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find

In [75]:
df_ords_dup = df_ords[df_ords.duplicated()]

In [77]:
df_ords_dup

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order, first_order]
Index: []

In [109]:
# The dataframe created to check duplicates has returned empty, whcih means an absence of duplicates in the data set.

## 8. Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice

In [110]:
df_prods_clean_no_dups.describe() # returns a summary of statistics for each numerical column in the DataFrame

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [82]:
# Let's look at the maximum value in the 'prices' column, which is unusually high '99999'.
# It's likely that such an extreme outlier is either a placeholder or and error.
# The 'product_id' column has a maximum value of 49688, assuming the count of 49672 is reflective of 49672 unique products there may be an error and/or duplicate values.

In [84]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_cleaned.csv'))

In [85]:
# Check number of unique product ids
df_unique_product_ids = df_prods_clean_no_dups['product_id'].nunique()

In [87]:
# method used to count the number of unique product IDs, which is then printed using an f-string
print(f"Number of unique product IDs: {df_unique_product_ids}")

Number of unique product IDs: 49670


In [88]:
# Check if product_id values are sequential and without gaps
df_all_ids = set(range(1, df_prods_clean_no_dups['product_id'].max() + 1))

In [89]:
# Create a set of actual ids in the dataframe
df_actual_ids = set(df_prods_clean_no_dups['product_id'].unique())

In [90]:
# Find any missing ids
missing_ids = df_all_ids - df_actual_ids

In [107]:
if missing_ids:
    print(f"The missing product IDs are: {sorted(missing_ids)}")
else:
    print("No missing product IDs.")

The missing product IDs are: [34, 69, 116, 262, 525, 1511, 1780, 2240, 2586, 3159, 3230, 3736, 4283, 4790, 6799, 26519, 38183, 40440]


In [ ]:
# These results indicate that some ids may be missing, the data collection process never captured them, or they may have been removed

In [ ]:
# The maximum value of 99999 also needs to be investigated 

In [92]:
outliers = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

In [113]:
print(outliers) # print the outliers

       product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0


In [95]:
# Now that we've determined where the outlier is located, there are several approaches that can be taken
# In this case, since the price skews that data immensely and it's clear that its an error and not an expensive product, we can remove it
df_prods_extra_clean_no_dups = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] != 99999]

In [111]:
df_prods_extra_clean_no_dups.describe() # returns a summary of statistics for each numerical column in the DataFrame

product_id      aisle_id  department_id        prices
count  49671.000000  49671.000000   49671.000000  49671.000000
mean   24850.172334     67.762115      11.728856      7.980256
std    14340.795118     38.316100       5.850806     66.952504
min        1.000000      1.000000       1.000000      1.000000
25%    12432.500000     35.000000       7.000000      4.100000
50%    24850.000000     69.000000      13.000000      7.100000
75%    37268.500000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  14900.000000

In [97]:
# It seems like there's yet another outlier. As such, we'll investigate the rows with the highest prices to see the other outliers
df_prods_sorted = df_prods_extra_clean_no_dups.sort_values(by='prices', ascending=False)

In [98]:
top_priced_products = df_prods_sorted.head() # This sorts the DataFrame df_prods by the ‘Price’ column in descending order, so the most expensive products are at the top. The head() method yields 5 rows by default.

In [114]:
print(top_priced_products) # print the results of above

       product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
9020         9020  Boneless Skinless Chicken Thighs        35             12   
40490       40486                   Chicken Tenders        49             12   
21468       21467            Wild Caught Raw Shrimp        15             12   
25580       25579     Naturally Smoked Trout Fillet        15             12   

        prices  
21554  14900.0  
9020      25.0  
40490     25.0  
21468     25.0  
25580     25.0  


In [100]:
# Now we know that the product priced 14900 is the only other outlier that we must remove.
df_prods_extra_clean_no_dups_2 = df_prods_extra_clean_no_dups[df_prods_extra_clean_no_dups['prices'] != 14900]

In [112]:
df_prods_extra_clean_no_dups_2.describe() # returns a summary of statistics for each numerical column in the DataFrame

product_id      aisle_id  department_id        prices
count  49670.000000  49670.000000   49670.000000  49670.000000
mean   24850.238716     67.761305      11.728770      7.680437
std    14340.931850     38.316060       5.850834      4.199381
min        1.000000      1.000000       1.000000      1.000000
25%    12432.250000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.750000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

#### The descriptive statistics now appear more normal and no visible errors are present

## 9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names

In [103]:
df_prods_extra_clean_no_dups_2.to_csv(os.path.join(path, 'Prepared Data', 'products_rechecked.csv'))

## 10. Save your Jupyter notebook and submit it here for your tutor to review

#### Done